lib/posix/pthread.c

```diff
@@ -151,73 +282,83 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
+	struct posix_thread *t = NULL;
 	const struct pthread_attr *attr = (const struct pthread_attr *)_attr;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid(attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
-		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
+	key = k_spin_lock(&q_lock);
+	if (!sys_dlist_is_empty(&ready_q)) {
+		/* spawn thread 't' directly from ready_q */
+		t = CONTAINER_OF(sys_dlist_get(&ready_q), struct posix_thread, q_node);
+	} else {
+		SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+			if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
+				/* thread has not been joined yet */
+				continue;
+			}
+
+			/* spawn thread 't' from done_q */
+			sys_dlist_remove(&t->q_node);
 			break;
 		}
 	}
-	k_spin_unlock(&pthread_pool_lock, key);
 
-	if (pthread_num >= CONFIG_MAX_PTHREAD_COUNT) {
-		return EAGAIN;
+	if (t != NULL) {
+		/* initialize thread state */
+		sys_dlist_append(&run_q, &t->q_node);
+		t->qid = POSIX_THREAD_RUN_Q;
+		t->detachstate = attr->detachstate;
+		if ((BIT(_PTHREAD_CANCEL_POS) & attr->flags) != 0) {
+			t->cancel_state = PTHREAD_CANCEL_ENABLE;
+		}
+		t->cancel_pending = false;
+		sys_slist_init(&t->key_list);
 	}
+	k_spin_unlock(&q_lock, key);
 
-	rv = pthread_mutex_init(&thread->state_lock, NULL);
-	if (rv != 0) {
-		key = k_spin_lock(&pthread_pool_lock);
-		thread->state = PTHREAD_EXITED;
-		k_spin_unlock(&pthread_pool_lock, key);
-		return rv;
+	if (t == NULL) {
+		/* no threads are ready */
+		return EAGAIN;
 	}
 
-	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+	err = pthread_barrier_init(&barrier, NULL, 2);
+	if (err != 0) {
+		/* cannot allocate barrier. move thread back to ready_q */
+		key = k_spin_lock(&q_lock);
+		sys_dlist_remove(&t->q_node);
+		sys_dlist_append(&ready_q, &t->q_node);
+		t->qid = POSIX_THREAD_READY_Q;
+		k_spin_unlock(&q_lock, key);
+		return EAGAIN;
+	}
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_state = (1 << _PTHREAD_CANCEL_POS) & attr->flags;
-	thread->cancel_pending = 0;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
+	/* spawn the thread and then wait for the spawned thread to cross our barrier */
+	k_thread_create(&t->thread, attr->stack, attr->stacksize, zephyr_thread_wrapper,
+			(void *)arg, UINT_TO_POINTER(barrier), threadroutine,
+			posix_to_zephyr_priority(attr->priority, attr->schedpolicy),
+			(~K_ESSENTIAL & attr->flags), K_MSEC(attr->delayedstart));
```

Comment:
I'm not sure - that's how it was done with the original implementation, so i didn't want to veer off too far.

Reply:
Oh wait - I guess the flags may be relevant to Zephyr but specifically `K_ESSENTIAL` wouldn't work because then we wouldn't be able to abort the thread (I guess that would trigger a panic (see [z_thread_abort()](https://github.com/zephyrproject-rtos/zephyr/blob/028e0a3c8d75fc5e2027f54dbb5a55aadbb1c490/kernel/sched.c#L1736))

Reply:
I added a check in `pthread_attr_is_valid()`